presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/PayloadJoinOptimizer.java

```diff
@@ -419,51 +493,67 @@ private static RowExpression zeroForType(Type type)
     private static class JoinContext
     {
         private final Set<VariableReferenceExpression> joinKeys = new HashSet<>();
-        Map<VariableReferenceExpression, VariableReferenceExpression> joinKeyMap;
+        private Map<VariableReferenceExpression, VariableReferenceExpression> joinKeyMap;
+        private Map<VariableReferenceExpression, RowExpression> projectionsToPush = new HashMap<>();
+
         int numJoins;
+        PlanNode payloadNode;
+
+        public JoinContext() {}
 
-        public PlanNode getFilterScanNode()
+        public Set<VariableReferenceExpression> getJoinKeys()
         {
-            return filterScanNode;
+            return joinKeys;
         }
 
-        public void setFilterScanNode(PlanNode planNode)
+        public void addKeys(ImmutableSet<VariableReferenceExpression> keys)
         {
-            this.filterScanNode = planNode;
+            joinKeys.addAll(keys);
         }
 
-        PlanNode filterScanNode;
+        public Map<VariableReferenceExpression, RowExpression> getProjectionsToPush()
+        {
+            return projectionsToPush;
+        }
 
-        public JoinContext() {}
+        public void addProjectionsToPush(Map<VariableReferenceExpression, RowExpression> map)
+        {
+            projectionsToPush.putAll(map);
+        }
 
-        public void addKeys(ImmutableSet<VariableReferenceExpression> keys)
+        public Map<VariableReferenceExpression, VariableReferenceExpression> getJoinKeyMap()
         {
-            joinKeys.addAll(keys);
+            return joinKeyMap;
         }
 
         public void setJoinKeyMap(Map<VariableReferenceExpression, VariableReferenceExpression> map)
         {
             joinKeyMap = map;
         }
 
-        public void incrementNumJoins()
+        public PlanNode getPayloadNode()
         {
-            numJoins++;
+            return payloadNode;
```

Comment:
Might be better to move the not null check here

Reply:
having null is valid, it's up to the caller what to do with that information